### DISCUSSION TITLE
[Solved] Combining two pictures and keeping the size of the largest one

### DISCUSSION THREAD

**Discussion by Tuttu** [ImageMagick/ImageMagick#2049] [Help] [answered]
Hello, 

Let's take the two following pictures: 
- PictureA.png - Size of 800x600 pixels. Part of it is transparent
- PictureB.png - Size of 300x150 pixels. 

I want to combine the two pictures in a single one. PictureA must be on top of PictureB. Part PictureB will be visible from the transparent area of PictureA. The final picture should have the size of PictureA.

I am using the following command : 
`magick.exe "composite -gravity East PictureA.png PictureB.png PictureA.png"`

The result is not what I'm expecting. PictureA is on top of PictureB but the resulting picture is the size of PictureB rather than PictureA. I have tried adding some parameters like -size or -resize but I was unable to get the result I wanted.

What should I do top get the result I want?

**Comment by snibgo** [comment, marked as answer]
What version of IM, on what platform? Please supply sample images.

What are those quotes doing? Remove them.

I suggest you use "magick", not "magick composite".

`magick PictureA.png PictureB.png -compose SrcOver -composite out.png`

Does that do what you want?

**Reply by Tuttu** [reply]
I am using `ImageMagick 7.0.10-2 Q16 x64` on Windows 10.
The quotes came from my original command line copy-paste. I'm using PowerShell to start ImageMagick and I forgot to clean the quotes when pasting the code here. :)

As for your suggestion, it almost did the trick. I had to use `-compose DstOver` to get the result I wanted. Thanks a lot for your help. Now, I will try to use `magick` alone more often. :)

**Comment by LorenzMichels** [comment]
I tried this with the following command: magick image.jpg overlay.png -compose SrcOver -composite result.png
I am on Windows 11 and use Version: ImageMagick 7.1.0-62 Q16-HDRI x64 32ce406:20230212

The overlay image is put on top of the image.jpg, but it isn't resized. In my case the overlay is way smaller than my images. I would like the overlay to get the same width of the image.jpg (and keep the aspect ration of the overlay). The overlay has to be centered (horizontally as well as vertically) as well.

Can you help me with that?  I have spent hours searching and tried several things, but I can't get it to work.

**Comment by snibgo** [comment]
> ... but it isn't resized.

IM won't resize images unless you tell it to. A common way to do this is with "`-resize WxH`" where W and H are numbers or expressions for the width and height. If the operation isn't decorated with "`!`", it will be resized to fit within a box of those dimensions, without changing the aspect ratio.

We can use the dimensions of the first image, like this:
```
magick image.jpg -set option:MYSIZE %wx%h ( overlay.png -resize %[MYSIZE] ) -compose SrcOver -composite result.png
```
If you do this in a BAT script, double every `%`.

**Reply by LorenzMichels** [reply]
Wauw, THANKS! Earlier this day I was very near this solution, but I didn't know I had to double the %-signs. I added the -gravity center option to set the overlay in the middle of the image. The completed batch-file now has this line:

magick image.jpg -set option:MYSIZE %%wx%%h ( overlay.png -resize %%[MYSIZE] ) -gravity center -compose SrcOver -composite result.jpg

Thanks again, this works like a charm.

**Comment by LorenzMichels** [comment]
Just in case someone else wants to do the same, here is what I have created today (thanks to snibgo). I wanted to make an export from photos with Lightroom Classic (LRC). LRC has possibilities to add textual or graphical watermarks. And of course I want to have both and don't want to do 2 exports manually. That's where ImageMagic (IM) came in. The created scripts can be used with/by the Export command from LRC (script is at the bottom).

I created two PNG files. One with a logo (Overlay - Logo.png) and the other with copyright text (Overlay - Copyright). The first I wanted to spread the whole image and the second I wanted to be in the lower right corner.  I created three command-files (windows batch files) being "Overlay - Copyright.cmd", "Overlay - Logo.cmd",  "Overlay - Logo and Copyright.cmd". I put these files in  "C:\Users\<username>\AppData\Roaming\Adobe\Lightroom\Export Actions".

If you've done so, you'll have to go to the Export dialogue of LRC and at the very bottom in the section Post-processing in After Export you have to select "Go to Export Actions Folder Now". Once you've done that all command files and programs you've placed in the Export Actions folder will become an option in the same drop-down menu you chose the last option from. The newly added options will carry the names you gave to the command files. You have to do this only once.

When all is in place, the following will happen using the scripts I made. In the example I will use the one which will add both. In LRC you choose Export and set the options you like. There is no need for watermarking, because the scripts will take care of that. In the After Export you choose for example "Overlay - Logo and Copyright". LRC will export the photos you have chosen in the directory you have set. When all the files are exported. The chosen script will 'fire'.

First the script will read the drive, path, filename and extension from the input file. These values will be put in separate variables. The drive and path is the location where LRC has put the exported files. The script will then make a sub-folder in that location with the name "Watermarked".

Then the two IM magic kicks in. At first the logo overlay will be resized to the width of the photo to be overlaid and it will be overlaid in the middel of the photo. Then the second IM statement will resize the copyright overlay to 0.3 times the height of the photo to be overlaid. This overlay will be put in the lower right corner.

I don't know whether LRC will send the exported photos one by one to my scripts or if it will send the complete list of files in one time. The script supports both methods. You can use the scripts manually as well by dropping a selection of photos on the script file (only one command windows will open). When exporting a lot of photos from LRC it will open multiple command windows (I think it uses one by one method) in which it will server multiple files. Exporting about 80 files resulted in 5 or 6 command windows. Wait for all the command windows will stop with the message "Press any key...". 

If LRC and the selected script are ready, you remain with the LRC exports without any watermarks and you will have the same photo's with watermark(s) in the Watermarked folder.

For some unknown reason, the script couldn't find my overlay files that are on the P-drive. That is strange, because the absolute path is hard-coded in the script. The script however looked in the path without drive letter (relatively). Strange! Therefore there is a line which first selects the P-drive.

The shift command will take care that all the files that are dropped on the script-file will be put in the %1 parameter and will loop throught the whole list. Below is the script for both overlays. the other two scripts are the same except for the ImageMagick lines. In each I deleted the line which wasn't needed for that script.

Hope this will help some other people.

```
@echo off

if [%1]==[] goto :eof
:loop

echo %1

set filedrive=%~d1
set filepath=%~p1
set filename=%~n1
set fileextension=%~x1

if not exist "%filedrive%%filepath%Watermarked" (
	md "%filedrive%%filepath%Watermarked"
)

P:

"C:\Program Files\ImageMagick-7.1.0-Q16-HDRI\magick.exe" %1 -set option:MYSIZE %%wx%%h ( "‪P:\Diversen\_LM logo font (!)\Overlay - Logo.png" -resize %%[MYSIZE] ) -gravity center -compose SrcOver -composite "%filedrive%%filepath%Watermarked\_%filename%%fileextension%" 

"C:\Program Files\ImageMagick-7.1.0-Q16-HDRI\magick.exe" "%filedrive%%filepath%Watermarked\_%filename%%fileextension%" -set option:SIZE2 %%[fx:int(h*0.3)] ( "‪P:\Diversen\_LM logo font (!)\Overlay - Copyright.png" -resize %%[SIZE2] ) -gravity southeast -compose SrcOver -composite -geometry +0+50 "%filedrive%%filepath%Watermarked\_%filename%%fileextension%" 

shift
if not [%1]==[] goto loop

:eof
pause
%SystemRoot%\explorer.exe  "%filedrive%%filepath%Watermarked\"
```
